### PR TITLE
#13118 #15250 Add highlighting

### LINKED ISSUE #13118: [ERD] Highlight selected FK and referenced tables with distinctive color on FK selection
For the diagram to be more useful please add color marking for the highlighted foreign key
This way all connecting lines and all instances of the same key would be highlighted/(marked with color) and it would be  visible on the diagram.

### LINKED ISSUE #15250: Add an ability to cycle through found object in ERD search
**Is your feature request related to a problem? Please describe.**
The search function in ERD moves focus to only one of the objects that is found, but doesn't allow to move the focus to another

**Describe the solution you'd like**
Add a button like in a simple search (Find Next) that will advance the focus to the next found object

dbeaver_vhAyDEH9fa

### PR DESCRIPTION
**Pull request by E1izabeth**
Highlighting was added to
- mark search results
- highlight association and PK on FK selection
- highlight PK and FK on association selection

Also jump to next/prev search result in endless loop is supported now.

https://user-images.githubusercontent.com/28875055/166811314-58e459ee-4456-40fa-ae97-0f8856d753fd.mp4

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.erd.ui/plugin.xml [hunk 1]
```diff
@@ -226,6 +226,20 @@
                 id="org.jkiss.dbeaver.erd.diagram.attributes.foreground"
                 value="COLOR_LIST_FOREGROUND">
             <description>%colorDefinition.org.jkiss.dbeaver.erd.diagram.attributes.foreground.description</description>
+        </colorDefinition>
+                <colorDefinition
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/plugin.xml]: Excess indentation.

File: plugins/org.jkiss.dbeaver.erd.ui/plugin.xml [hunk 2]
```diff
@@ -226,6 +226,20 @@
...
+                label="%colorDefinition.org.jkiss.dbeaver.erd.diagram.search.highlighting.label"
+                categoryId="org.jkiss.dbeaver.ui.presentation.erd"
+                id="org.jkiss.dbeaver.erd.diagram.search.highlighting"
+                value="255,222,0">
+            <description>%colorDefinition.org.jkiss.dbeaver.erd.diagram.search.highlighting.description</description>
+        </colorDefinition>
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/plugin.xml:236]: Excess indentation.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java [hunk 3]
```diff
@@ -183,6 +187,10 @@ public ERDDecorator getDecorator() {
         }
         return decorator;
     }
+    
+    public ERDHighlightingManger getHighlightingManager() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java]: Please specify `@NotNull`/`@Nullable` annotations for the method.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java [hunk 4]
```diff
@@ -1325,6 +1380,11 @@ public void cancelSearch() {
                 curSearchPattern = null;
                 if (resultsFound) {
                     resultsFound = false;
+                    results = null;
+                    currentItem = null;
+                    resultsIterator = null;
+                    highlightings.forEach(h -> h.release());
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java]:
```suggestion
                    highlightings.forEach(ERDHighlightingHandle::release);
```

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    @NotNull
+    private final Map<IFigure, PartHighlighter> highlightedParts = new HashMap<>();
+
+    private static final class HighlightingEntry {
+        public final Color color;
+
+        public HighlightingEntry(@NotNull Color color) {
+            this.color = color;
+        }
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java]: I'm not sure what the point of this wrapper class is.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                    }
+                    part.setOpaque(true);
+                }
+                part.repaint();
+            } catch (Throwable ex) {
+                ex.printStackTrace();
+            }
+        }
+        
+        public ERDHighlightingHandle highlight(@NotNull Color color) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java]: Please specify `@NotNull`/`@Nullable` annotations for the method.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+import org.eclipse.draw2dl.Connection;
+import org.eclipse.draw2dl.IFigure;
+import org.eclipse.swt.graphics.Color;
+import org.jkiss.code.NotNull;
+import org.jkiss.dbeaver.erd.model.ERDEntityAttribute;
+import org.jkiss.dbeaver.erd.ui.part.AssociationPart;
+import org.jkiss.dbeaver.erd.ui.part.AttributePart;
+import org.jkiss.dbeaver.erd.ui.part.EntityPart;
+import org.jkiss.dbeaver.model.sql.parser.ListNode;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java]: This is a big no-no. ERD has no dependency on the SQL parser and it should not.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                    if (part instanceof Connection) {
+                        part.setForegroundColor(highlightings.getLast().color);
+                    } else {
+                        part.setBackgroundColor(highlightings.getLast().color);
+                    }
+                    part.setOpaque(true);
+                }
+                part.repaint();
+            } catch (Throwable ex) {
+                ex.printStackTrace();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java]: Please log the exception instead of dumping its stack trace.

Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java]: What are the conditions for an exception to be thrown in the first place? If it's only related to the `repaint` method, then please shrink the `try-catch` body to cover that one method.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        }
+        
+    }
+
+    @NotNull
+    public ERDHighlightingHandle highlight(@NotNull IFigure part, @NotNull Color color) {
+        return highlightedParts.computeIfAbsent(part, p -> new PartHighlighter(p)).highlight(color);
+    }
+    
+    private ERDHighlightingHandle makeHighlightingGroupHandle(ListNode<ERDHighlightingHandle> highlightings) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java]: Please specify `@NotNull`/`@Nullable` annotations for the method.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        } else {
+            return () -> {
+                for (ERDHighlightingHandle highlighting: highlightings) {
+                    highlighting.release();
+                }
+            };
+        }
+    }
+    
+    public ERDHighlightingHandle highlightAttributeAssociations(AttributePart attributePart, Color color) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java]: Please specify `@NotNull`/`@Nullable` annotations for the method.

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/controls/ProgressPageControl.java [hunk 11]
```diff
@@ -387,6 +389,7 @@ public void keyPressed(KeyEvent e) {
                             cancelSearch(true);
                             break;
                         case SWT.CR:
+                            performSearch(SearchType.NEXT);
                         case SWT.ARROW_UP:
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/controls/ProgressPageControl.java:394]: Please add `/* fall-through */` comment.

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/controls/ProgressPageControl.java [hunk 12]
```diff
@@ -419,20 +422,16 @@ public IStatus runInUIThread(IProgressMonitor monitor) {
             //ToolBar searchTools = new ToolBar(searchControlsComposite, SWT.HORIZONTAL);
             if (searchToolbarManager == null) {
                 searchToolbarManager = new ToolBarManager(SWT.FLAT | SWT.HORIZONTAL);
-                // Do not add prev/next buttons - they doesn't make sense now.
-                // Keep code just in case
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/controls/ProgressPageControl.java:423]: That's cool that this part happened to be helpful.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+import org.eclipse.swt.graphics.Color;
+import org.jkiss.code.NotNull;
+import org.jkiss.dbeaver.erd.model.ERDEntityAttribute;
+import org.jkiss.dbeaver.erd.ui.part.AssociationPart;
+import org.jkiss.dbeaver.erd.ui.part.AttributePart;
+import org.jkiss.dbeaver.erd.ui.part.EntityPart;
+import org.jkiss.dbeaver.model.sql.parser.ListNode;
+
+
+public class ERDHighlightingManger {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java]: Misspelled `Manager`

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java (hunk 7 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManger.java]: It already has this dependency. I have just reused utility class.

File: plugins/org.jkiss.dbeaver.ui/META-INF/MANIFEST.MF [hunk 14]
```diff
@@ -37,6 +37,7 @@ Require-Bundle: org.eclipse.core.runtime,
  org.eclipse.e4.ui.css.swt.theme,
  org.eclipse.e4.ui.workbench.renderers.swt,
  org.jkiss.dbeaver.model,
- org.eclipse.ui.forms
+ org.eclipse.ui.forms,
+ org.eclipse.ui.workbench.texteditor
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui/META-INF/MANIFEST.MF:41]: Why is this bundle required?

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AttributePart.java [hunk 15]
```diff
@@ -117,6 +124,16 @@ public void setSelected(int value) {
             }
         }
         columnLabel.repaint();
+
+        if (value != EditPart.SELECTED_NONE) {
+            if (this.getViewer() instanceof ERDGraphicalViewer && associatedRelationsHighlighing == null) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AttributePart.java:129]: These `if` statements can be merged.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AssociationPart.java [hunk 16]
```diff
@@ -226,34 +231,15 @@ public void setSelected(int value) {
...
-        }
-    }
-
-    private List<AttributePart> getEntityAttributes(EntityPart source, List<ERDEntityAttribute> columns) {
-        List<AttributePart> result = new ArrayList<>();
-        for (Object attrPart : source.getChildren()) {
-            if (attrPart instanceof AttributePart && columns.contains(((AttributePart)attrPart).getAttribute())) {
-                result.add((AttributePart)attrPart);
+        if (value != EditPart.SELECTED_NONE) {
+            if (this.getViewer() instanceof ERDGraphicalViewer && associatedAttributesHighlighing == null) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AssociationPart.java:235]: These `if` statements can be merged.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManager.java [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+        
+        return highlightings;
+    }
+
+    @NotNull
+    private List<AttributePart> getEntityAttributes(@NotNull EntityPart source, @NotNull List<ERDEntityAttribute> columns) {
+        List<AttributePart> result = new ArrayList<>(columns.size());
+        for (Object attrPart : source.getChildren()) {
+            if (attrPart instanceof AttributePart && columns.contains(((AttributePart)attrPart).getAttribute())) {
+                result.add((AttributePart)attrPart);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManager.java]:
```suggestion
            if (attrPart instanceof AttributePart && columns.contains(((AttributePart) attrPart).getAttribute())) {
                result.add((AttributePart) attrPart);
```

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManager.java [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+
+        @NotNull
+        public ERDHighlightingHandle highlight(@NotNull Color color) {
+            HighlightingEntry entry = new HighlightingEntry(color);
+            highlightings.addLast(entry);
+            this.refresh();
+            
+            return () -> {
+                // we are removing entry somewhere in the middle of the highlighting stack, so reference identity is required,
+                // which Color class does not follow, as it supports by-value equivalence
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManager.java]: This comment will make more sense if placed directly near the `HighlightingEntry` class.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java [hunk 19]
```diff
@@ -1267,56 +1276,107 @@ private boolean restoreVisualSettings(DiagramPart oldDiagram, EntityDiagram newD
...
+                    // direction change gets current item again as if it's a new loop initialization 
+                    jumpToNext(isFindNext); 
+                }
+                isPrevStepWasFwd = isFindNext;
+                jumpToNext(isFindNext);
+
+                ERDGraphicalViewer graphicalViewer = getGraphicalViewer();
+                graphicalViewer.deselectAll();
+                graphicalViewer.select((EditPart)currentItem);
+                graphicalViewer.reveal((EditPart)currentItem);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java]:
```suggestion
                graphicalViewer.select((EditPart) currentItem);
                graphicalViewer.reveal((EditPart) currentItem);
```

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java [hunk 20]
```diff
@@ -1267,56 +1276,107 @@ private boolean restoreVisualSettings(DiagramPart oldDiagram, EntityDiagram newD
...
+                    Object obj = nodes.get(0);
+                    if (obj instanceof DBPNamedObject && obj instanceof EditPart) {
+                        Color color = UIUtils.getColorRegistry().get(ERDUIConstants.COLOR_ERD_SEARCH_HIGHLIGHTING);
+                        for (Object node : nodes) {
+                            DBPNamedObject erdNode = (DBPNamedObject) node;
+                            if (matchesSearch(erdNode)) {
+                                resultsFound = true;
+                                results.add(erdNode);
+                                if (erdNode instanceof GraphicalEditPart) {
+                                    highlightings.add(highlightingManager.highlight(((GraphicalEditPart)erdNode).getFigure(), color));
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java]:
```suggestion
                                    highlightings.add(highlightingManager.highlight(((GraphicalEditPart) erdNode).getFigure(), color));
```

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java [hunk 21]
```diff
@@ -116,7 +119,7 @@
  */
 public abstract class ERDEditorPart extends GraphicalEditorWithFlyoutPalette
         implements DBPDataSourceTask, IDatabaseModellerEditor, ISearchContextProvider, IRefreshablePart, INavigatorModelView {
-    private static final Log log = Log.getLog(Searcher.class);
+    private static final Log searcherLog = Log.getLog(Searcher.class);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java:119]: Why this change?

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java (hunk 21 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java:119]: Because it's log for Searcher class which is placed in the parent class. I assume it would be better to highlight it to prevent accidentally usage in parent class.

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java:119]: I see. Is it possible to move it into the `Searcher` class?

File: plugins/org.jkiss.dbeaver.ui/META-INF/MANIFEST.MF (hunk 14 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui/META-INF/MANIFEST.MF:41]: We need it to use constants to check search type such as SEARCH_NEXT, SEARCH_PREVIOUS

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java (hunk 21 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java:119]: Unfortunately, it's not a thing until Java 16. That's sad.

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java:119]: It's not possible due to java language level.

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java:119]: static declarations in inner class are not supported yet

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AttributePart.java (hunk 15 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AttributePart.java:129]: These two conditions should not be merged, it'll break the logic.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AssociationPart.java (hunk 16 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AssociationPart.java:235]: These two conditions should not be merged, it'll break the logic.

File: plugins/org.jkiss.dbeaver.ui/META-INF/MANIFEST.MF (hunk 14 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.ui/META-INF/MANIFEST.MF:41]: Can it be added to the `org.jkiss.dbeaver.erd.ui` instead?

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui/META-INF/MANIFEST.MF:41]: Constants from `IWorkbenchActionDefinitionIds` are used in `ProgressPageControl` which comes from `org.jkiss.dbeaver.ui.controls` package. So, no, it can't.

**Comment by Matvey16** [comment]
Focus can't be returned to tables after it goes to columns
dbeaver_fLRYYxNUMu

**Comment by Matvey16** [comment]
Verified
